pkg/client/apiutil/filterdiscoveryclient.go

```diff
@@ -0,0 +1,104 @@
+package apiutil
+
+import (
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+)
+
+// defaultRetries is the number of times a resource discovery is repeated if an api group disappears on the fly (e.g. ThirdPartyResources).
+const defaultRetries = 2
+
+type FilterDiscoveryClient struct {
+	Discovery discovery.DiscoveryInterface
+	Filter    func(groupVersion string) bool
+}
+
+// ServerResourcesForGroupVersion is doing the actual filtering
+func (d *FilterDiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	if d.Filter(groupVersion) {
+		return d.Discovery.ServerResourcesForGroupVersion(groupVersion)
+	}
+	return &metav1.APIResourceList{
+		GroupVersion: groupVersion,
+		APIResources: []metav1.APIResource{},
+	}, nil
+}
+
+// ServerGroupsAndResources calls the public module function with d as the interface
+// so that it will callback to d.ServerResourcesForGroupVersion(groupVersion) for filtering
+func (d *FilterDiscoveryClient) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	return withRetries(defaultRetries, func() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
```

Comment:
add a note here and below that this is copied from the client-go discovery client, in case we're wondering what's going on here.

Reply:
do we even needs the retries, since this is being used in the dynamic restmapper, which'll retry anyway?

Reply:
We needed to copy the retries logic only for two cases.
For most of the wrapped functions we just call the underlying discovery interface directly and therefore do not need to add retries.
But for the two calls that we override to implement the filtering which are `ServerGroupsAndResources()` and `ServerPreferredResources()` I had to copy the retries logic too.

I certainly didn't want to do it, but also didn't want to leave these calls to fail too easily.

Other suggestions are certainly welcome.